src/monitoring/history.rs

```diff
@@ -0,0 +1,144 @@
+            diff.mean_time = if mean_time.is_nan() { 0.0 } else { mean_time }
+        }
+
+        self.storage.insert(snapshot_id.clone(), query_snapshot);
+        diff
+    }
+
+    fn make_storage_id(&mut self, snapshot: &Snapshot) -> String {
+        format!("{}{}", snapshot.dbid, snapshot.query)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use drivers::postgres::types;
+
+    #[test]
+    fn diff_calls_and_mean_time_is_corrected() {
+        let mut query_snapshot1 = self::get_base_snapshot();
+        query_snapshot1.calls = 5;
+        query_snapshot1.total_time = 10 as types::DoublePrecision;
+
+        let mut query_snapshot2 = self::get_base_snapshot();
+        query_snapshot2.calls = 7;
+        query_snapshot2.total_time = 20 as types::DoublePrecision;
+
+        let mut history = History::new();
+        history.save_stat_and_get_diff(query_snapshot1);
+
+        let diff: Diff = history.save_stat_and_get_diff(query_snapshot2);
+
+        assert!(0.2f64 - 0.001 <= diff.mean_time && diff.mean_time <= 0.2f64 + 0.001);
+        assert_eq!(2, diff.calls);
+    }
+
+    #[test]
+    fn identical_queries_from_different_db_will_not_be_compatible() {
+        let mut query_snapshot1 = self::get_base_snapshot();
+        query_snapshot1.dbid = 1;
+        query_snapshot1.query = String::from("SELECT 1");
+        query_snapshot1.calls = 5;
+        query_snapshot1.total_time = 10 as types::DoublePrecision;
+
+        let mut query_snapshot2 = self::get_base_snapshot();
+        query_snapshot2.dbid = 2;
+        query_snapshot2.query = String::from("SELECT 1");
+        query_snapshot2.calls = 7;
+        query_snapshot2.total_time = 20 as types::DoublePrecision;
+
+        let mut history = History::new();
+        history.save_stat_and_get_diff(query_snapshot1);
+
+        let diff: Diff = history.save_stat_and_get_diff(query_snapshot2);
+
+        assert_eq!(diff.calls, 0);
+        assert_eq!(diff.mean_time, 0.0);
+    }
+
+    #[test]
+    fn full_identical_queries_will_not_be_diff() {
+        let mut query_snapshot1 = self::get_base_snapshot();
+        query_snapshot1.dbid = 1;
+        query_snapshot1.query = String::from("SELECT 1");
+        query_snapshot1.calls = 5;
+        query_snapshot1.total_time = 10 as types::DoublePrecision;
+
+        let mut query_snapshot2 = self::get_base_snapshot();
+        query_snapshot2.dbid = 1;
+        query_snapshot2.query = String::from("SELECT 1");
+        query_snapshot2.calls = 5;
+        query_snapshot2.total_time = 10 as types::DoublePrecision;
+
+        let mut history = History::new();
+        history.save_stat_and_get_diff(query_snapshot1);
+
+        let diff: Diff = history.save_stat_and_get_diff(query_snapshot2);
+
+        println!("{}", diff.mean_time);
+
+        assert_eq!(diff.calls, 0);
+        assert_eq!(diff.mean_time, 0.0);
+    }
+
+    #[test]
+    fn save_stat_and_get_diff() {
```

Comment:
It seems this test tests nothing because no assertions here and panic is not possible